packages/core/src/types.ts

```diff
@@ -444,9 +444,7 @@ export function getElementDefinition(typeName: string, propertyName: string): El
  * @param value The object to check
  * @returns True if the input is of type 'object' and contains property 'resourceType'
  */
-export function isResource<T extends Resource = Resource>(
-  value: T | Reference<T> | string | undefined | null
-): value is T {
+export function isResource<T extends Resource = Resource>(value: unknown): value is T {
```

Comment:
This type signature didn't immediately make sense to me; is the generic type parameter necessary slash could this just be `export function isResource(value: unknown): value is Resource`?

Reply:
It can still be useful if you pass in a value with type `Patient | undefined`, then the type assertion will be `value is Patient` rather than `value is Resource`.

Reply:
Yep, good call.  Just didn't clean it up all the way.

Reply:
TypeScript is smart. 

```ts

interface Resource {
    resourceType: string;
}

interface Patient extends Resource {
    name: string;
}

interface Observation extends Resource {
    value: number;
}

function isResource1(input: unknown): input is Resource {
    return !!input && (typeof input === 'object') && ('resourceType' in input);
}

function isResource2<T extends Resource>(input: unknown): input is T {
    return !!input && (typeof input === 'object') && ('resourceType' in input);
}

const myInput = {} as Patient | undefined;

if (isResource1(myInput)) {
    console.log('1', myInput); // myInput is Patient, even though I assumed it would be Resource
}

if (isResource2(myInput)) {
    console.log('2', myInput); // myInput is Patient
}

```

Reply:
Hm, `readReference` is not as smart.  Requires changes like this:

```
-    if (isReference<T>(value)) {
-      return medplum.getCachedReference(value);
+    if (isReference(value)) {
+      return medplum.getCachedReference(value as Reference<T>);
```